### PR TITLE
EDGEAI-1289: Optimize CDR writer via Builder pattern

### PR DESCRIPTION
**Pull request by sebastient**
## Summary

Complete overhaul of the CDR write path for the 3.2.0 release: every buffer-backed message type gets a `Foo::builder()` API with three finalizers for allocation-free buffer reuse, every fixed-size field gets an in-place `set_*` mutator for the scalar fast path, every legacy `Foo::new(...)` is `#[deprecated(since = "3.2.0")]` for removal in 4.0, and a parallel C FFI builder surface (`ros_<type>_builder_*`) mirrors the Rust API for downstream C consumers.

**Stats:** 20 commits, +16635 / −36 across 21 files. 301 tests passing, 0 failing. Clippy `--all-targets -- -D warnings` clean, fmt clean.

### Rust Builder API (27 buffer-backed types)

`Foo::builder() -> FooBuilder<'a>` with:
- `build() -> Result<Foo<Vec<u8>>, CdrError>` — allocates fresh buffer (drop-in for legacy `new()`)
- `encode_into_vec(&mut Vec<u8>) -> Result<(), CdrError>` — reuses caller's Vec (grows on first call, no re-alloc on subsequent equal-size encodes)
- `encode_into_slice(&mut [u8]) -> Result<usize, CdrError>` — fixed-capacity encode; errors `BufferTooShort` without mutating on error

Setters are `&mut self -> &mut Self`, supporting both chained one-shot and named-reuse idioms (same auto-ref mechanism as `std::process::Command`).

Per-field storage: strings as `Cow<'a, str>`, bulk data as `&'a [u8]`, nested views as `&'a [View<'a>]`, fixed scalars and `CdrFixed` types by value.

**Messages covered:**
- std_msgs: `Header`
- foxglove_msgs: `FoxgloveCompressedVideo`, `FoxgloveTextAnnotation`, `FoxglovePointAnnotation`, `FoxgloveImageAnnotation`
- sensor_msgs: `CompressedImage`, `Image`, `Imu`, `NavSatFix`, `PointField`, `PointCloud2`, `CameraInfo`, `MagneticField`, `FluidPressure`, `Temperature`, `BatteryState`
- edgefirst_msgs: `Mask`, `LocalTime`, `RadarCube`, `RadarInfo`, `Track`, `DetectBox`, `Detect`, `CameraFrame`, `Model`, `ModelInfo`, `Vibration`

`DmaBuffer` excluded — already `#[deprecated]` for 4.0 removal.

### In-place scalar setter audit (108 new `set_*` methods)

Every fixed-size field on every buffer-backed message now has a `set_*` method that writes directly into the CDR buffer without re-serialising. For a 2 MB camera frame where only the timestamp changes, `img.set_stamp(now())` is ~8 byte writes vs. the builder's full re-encode — a ~5 orders-of-magnitude win on the scalar-only hot path.

Setters respect CDR alignment (including EDGEAI-1243's position-dependent `NavSatStatus` case) and use new `wr_f64`, `wr_u64`, `wr_i16`, `wr_u16`, `wr_bool`, `wr_f32`, `wr_i8` helpers in `src/cdr.rs`.

### Deprecation

- Every legacy `Foo::new(...)` marked `#[deprecated(since = "3.2.0", note = "use Foo::builder() for allocation-free buffer reuse; Foo::new will be removed in 4.0")]`.
- Legacy `ros_*_encode(...)` C FFI one-shot functions kept for source compatibility but all internal callers migrated to builder; documented as deprecated in `CAPI.md` with 4.0 removal scheduled.
- Internal FFI encoders (`ros_header_encode`, `ros_image_encode`, `ros_compressed_image_encode`, `ros_compressed_video_encode`, `ros_mask_encode`) migrated to builder — no C-API behavioural change.

### C FFI builder surface (298 new `extern "C"` functions)

Every Rust builder has a parallel C handle-based API:
- `ros_<type>_builder_new()` / `_free()` — opaque handle
- `ros_<type>_builder_set_<field>(b, ...)` — strings copy internally, bulk data borrows (caller keeps alive until next setter/build/encode_into/free)
- `ros_<type>_builder_build(b, **out, *out_len)` — allocate fresh CDR bytes
- `ros_<type>_builder_encode_into(b, *buf, cap, *out_len)` — encode into caller-owned buffer; EINVAL / ENOBUFS / EBADMSG error convention

CdrFixed struct fields (`Vector3`, `Quaternion`, `NavSatStatus`, `RegionOfInterest`, `FoxglovePoint2`, `FoxgloveColor`) exposed as primitive args — no C structs across the ABI.

Nested view sequences (`CameraFrame::planes`, `Detect::boxes`, `Model::boxes`/`masks`, `PointCloud2::fields`, `FoxgloveImageAnnotation::{circles,points,texts}`) exposed via `ros_<elem>_elem_t` C-POD descriptors with borrowed pointers; Rust side materialises `Vec<View<'_>>` at build time.

### Version + docs

- `Cargo.toml`, `edgefirst_msgs/package.xml`, `edgefirst/schemas/__init__.py` bumped to **3.2.0**
- `CHANGELOG.md` covers the Added/Deprecated/Migration sections
- `ARCHITECTURE.md` updated with final design (no Phase 1/Phase 2 framing), scalar fast-path guidance
- `CAPI.md` updated with builder as primary pattern; legacy `ros_*_encode` marked deprecated; entry-point table expanded

## Test coverage

**301 tests, 0 failures, 3 doc-test ignored (pre-existing).**

- **158 lib tests** (including setter parity guard test covering every new scalar setter)
- **27 builder byte-parity tests** across 4 integration test files (`builder_parity_std_msgs.rs`, `builder_parity_foxglove.rs`, `builder_parity_sensor_msgs.rs`, `builder_parity_edgefirst.rs`, plus the 3 in `cdr_golden.rs`) — every builder verified to produce bit-identical bytes vs. legacy `new()`
- **27 C FFI smoke tests** (`builder_ffi_smoke.rs`) — every `ros_<type>_builder_*` verified to produce the same bytes as the Rust builder
- **10 edgefirst_msgs lib tests**, **4 foxglove lib tests**, **9 sensor_msgs lib tests**, **1 std_msgs lib test**
- **7 builder reuse / slice-bounds tests** (`builder_reuse.rs`) — capacity stability, Vec shrinking, `BufferTooShort` canary-byte non-mutation
- **76 cross-language `cdr_golden` tests**, **4 mcap roundtrip tests**, **5 doc tests**

`cargo test` / `cargo clippy --all-targets -- -D warnings` / `cargo fmt --all -- --check` all clean.

## Why no in-place variable-length shuffle?

Considered and rejected: `set_encoding("longer")` memmove-ing the tail and patching every downstream offset is fragile for nested sequences (`CameraFrame::planes`, `Model::boxes`/`masks`, `Detect::boxes`, `FoxgloveImageAnnotation::*`) and position-dependent alignment (`NavSatStatus` in `NavSatFix`, per EDGEAI-1243). The builder re-serialises the whole buffer uniformly across all messages and reuses the same allocation the caller was already going to reuse.

## Explicitly deferred (out of scope for 3.x)

- **PyO3 bindings.** Today's Python bindings are hand-written dataclasses and stay that way through the 3.x series. The PyO3 migration is a separate architectural change tracked elsewhere.

## Commits

| SHA | Scope |
|---|---|
| `17bc7dd` | `ImageBuilder` + byte-parity test |
| `0470ce0` | `ImageBuilder` reuse / shrink / slice-bounds tests |
| `9d66f0c` | `CameraFrameBuilder` + byte-parity + validation-parity tests |
| `9b50b16` | `PointCloud2Builder` + byte-parity test |
| `90610dd` | Reuse tests extended to CameraFrame / PointCloud2 |
| `cb94a62` | ARCHITECTURE.md subsection |
| `af4e0ec` | `cargo fmt` |
| `145833a` | Deprecation markers on Image/CameraFrame/PointCloud2 `new()` |
| `6b61d76` | `HeaderBuilder` + deprecation + FFI migration |
| `8e2a7dc` | 4 Foxglove builders + deprecations + FFI migration |
| `95f1fc2` | 10 edgefirst_msgs builders + deprecations |
| `64eef05` | 9 sensor_msgs builders + deprecations + FFI migration |
| `e934a07` | Version bump to 3.2.0, CHANGELOG |
| `ab987c5` | ARCHITECTURE.md + CAPI.md for 3.2.0 |
| `a45c1b1` | In-place setter audit: foxglove (10 setters) |
| `954f77a` | In-place setter audit: edgefirst_msgs (44 setters) |
| `f8dbd9d` | In-place setter audit: sensor_msgs (54 setters) |
| `6100d16` | C FFI builders: Header, Image, FluidPressure, Temperature (34 fns) |
| `d145055` | C FFI builders: sensor_msgs remainder (99 fns) |
| `405e05f` | C FFI builders: edgefirst_msgs + foxglove_msgs (165 fns) |

## Related

- Ticket: **EDGEAI-1289**
- Parent: EDGEAI-1092 (CDR Optimizations umbrella)
- Motivating bug: EDGEAI-1243 (NavSatFix alignment — no-shuffle decision)

### CODE REVIEW

File: CAPI.md [hunk 1]
```diff
@@ -210,28 +210,81 @@ ros_image_free(img);
...
+| `ros_<type>_set_<field>(handle, ...)` | Write a fixed-size field in place (no re-serialisation) |
 | `ros_<type>_as_cdr(handle, &len)` | Borrow the raw CDR buffer (points into caller's data) |
 | `ros_<type>_free(handle)` | Release the handle (does NOT free the source data) |
-| `ros_<type>_encode(...)` | Construct CDR bytes from field values (allocates) |
+| `ros_<type>_builder_new()` | Create a builder handle |
+| `ros_<type>_builder_set_<field>(b, ...)` | Set a field on the builder (strings copy; bulk data borrows) |
+| `ros_<type>_builder_build(b, &bytes, &len)` | Allocate a fresh CDR buffer and encode |
+| `ros_<type>_builder_encode_into(b, buf, cap, &out_len)` | Encode into caller-owned buffer; errors if too small |
+| `ros_<type>_builder_free(b)` | Release the builder handle |
+| `ros_<type>_encode(...)` | **Deprecated** — one-shot encoder, removed in 4.0 |
```
[the remark this answers is not in the document]
Comment (sebastient) [reply CAPI.md:287]: Addressed in `b580609`. Added 120 `ros_<type>_set_<field>(buf, len, ...)` in-place setters mirroring the Rust `impl<B: AsRef<[u8]> + AsMut<[u8]>> Foo<B>::set_*` methods. These are stateless — each call re-parses the buffer via `from_cdr` to locate the field, then writes in place. Fixed-size fields only; variable-length fields require the builder API. CAPI.md updated (`ab987c5` and follow-ups) with explicit guidance on when to use each.

File: include/edgefirst/schemas.h [hunk 2]
```diff
@@ -684,6 +728,75 @@ int ros_image_encode(uint8_t** out_bytes, size_t* out_len,
...
+
+/** @brief Set the step (row stride in bytes) field. */
+void ros_image_builder_set_step(ros_image_builder_t* b, uint32_t v);
+
+/**
+ * @brief Set the data bulk byte sequence (BORROWED — must remain valid
+ *        until the next setter, build, encode_into, or free call).
+ */
+void ros_image_builder_set_data(ros_image_builder_t* b,
+                                const uint8_t* data, size_t len);
```
[the remark this answers is not in the document]
Comment (sebastient) [reply include/edgefirst/schemas.h:786]: Addressed in `2e2d12c`. `ros_<type>_builder_set_<data>` setters (25 of them) changed return type from `void` to `i32`; `(NULL, len>0)` now returns `-1` with `errno = EINVAL`. `(NULL, 0)` and `(ptr, 0)` remain valid as explicit clears. Doxygen updated. Smoke tests cover the error path.

File: include/edgefirst/schemas.h [hunk 3]
```diff
@@ -684,6 +728,75 @@ int ros_image_encode(uint8_t** out_bytes, size_t* out_len,
...
+
+/** @brief Set the stamp field. */
+void ros_image_builder_set_stamp(ros_image_builder_t* b,
+                                 int32_t sec, uint32_t nsec);
+
+/**
+ * @brief Set the frame_id field (string is copied into the builder).
+ * @return 0 on success, -1 on error (errno: EINVAL for NULL handle).
+ */
+int  ros_image_builder_set_frame_id(ros_image_builder_t* b, const char* s);
```
[the remark this answers is not in the document]
Comment (sebastient) [reply include/edgefirst/schemas.h:758]: Addressed in `2e2d12c`. All 44 `ros_<type>_builder_set_<str>` string setters now enforce: `NULL` pointer → `-1` with `EINVAL`, and invalid UTF-8 → `-1` with `EINVAL`. Shared `c_to_str_checked` helper introduced so the rule is uniform. Doxygen updated. Smoke tests cover NULL and non-UTF-8 paths.

File: include/edgefirst/schemas.h [hunk 4]
```diff
@@ -1480,6 +1795,161 @@ uint32_t ros_point_cloud2_get_fields_len(const ros_point_cloud2_t* view);
...
+ * NUL-terminated string alive (and the enclosing array) until the next
+ * `ros_point_cloud2_builder_set_fields` call on the builder, the next
+ * `build`/`encode_into`, or `free`.
+ */
+typedef struct ros_point_field_elem_s {
+    const char* name;
+    uint32_t    offset;
+    uint8_t     datatype;
+    uint32_t    count;
+} ros_point_field_elem_t;
```
[the remark this answers is not in the document]
Comment (sebastient) [reply include/edgefirst/schemas.h:1895]: Addressed in `2e2d12c`. Descriptor-struct string pointers (`ros_point_field_elem_t.name`, `ros_detect_box_elem_t.label`/`track_id`, `ros_mask_elem_t.encoding`, `ros_foxglove_text_annotation_elem_t.text`) now cause `build()` / `encode_into()` to return `-1` with `EINVAL` if any element in a bound array has a `NULL` string pointer. The `to_views`/`to_vec` helpers became fallible. Doxygen on each descriptor struct updated: "string pointers must be non-NULL valid NUL-terminated UTF-8; NULL causes build/encode to fail with EINVAL."

**Comment by sebastient** [comment]
PR review feedback addressed. Four new commits:

| SHA | Addresses |
|---|---|
| `b580609` | **C FFI in-place setters added.** 120 `ros_<type>_set_<field>(buf, len, ...)` functions mirroring Rust's 108 setters, exposing every fixed-size field on every buffer-backed type. Stateless: each call re-parses the CDR buffer via `from_cdr` to locate the field, then writes in place. Variable-length fields (strings, bulk data, nested sequences) are explicitly not exposed — those require the builder. |
| `2e2d12c` | **NULL / invalid-UTF-8 validation tightened.** 44 string setters return `-1` + `EINVAL` on NULL or non-UTF-8 input (was silently empty-string). 25 bulk-data setters (previously `void`) now return `i32`; `(NULL, len>0)` returns `-1` + `EINVAL`. 8 fixed-array setters (covariances, camera matrices) same treatment. Descriptor structs (`ros_point_field_elem_t`, `ros_detect_box_elem_t`, `ros_mask_elem_t`, `ros_foxglove_text_annotation_elem_t`) reject NULL string pointers at `build()` / `encode_into()` time. |
| `5681a76` | **LocalTime setter alignment fix.** Found during FFI smoke-test work: `set_date` / `set_time` / `set_timezone` computed positions from `offsets[0]` without CDR alignment, breaking wire-round-trip for odd-length `frame_id`. Fixed with `cdr_align`, verified with round-trip tests across all four `frame_id` parities. |

**CAPI.md** updated with explicit guidance (per your ask):

> **When to use in-place setters vs. the builder.** Fixed-size fields can be mutated via `ros_<type>_set_<field>(buf, len, ...)` — re-parses the buffer to locate the field, writes directly, no re-serialisation. Variable-length fields (strings, byte sequences, nested view sequences) **cannot** be mutated in place because changing their length would shift every subsequent field. Use the builder API for those.
>
> Rule of thumb: if the field's wire size can change, use the builder. Otherwise, use the in-place setter.

**Test coverage:**
- 336 tests passing, 0 failing (+35 since last review)
- New `tests/in_place_setters_smoke.rs` — 29 tests covering FFI in-place setters (one per message type) + 3 error-path tests (NULL buf, malformed buf, NULL descriptor)
- Extended `tests/builder_ffi_smoke.rs` — +6 error-path tests (NULL `c_char`, non-UTF-8 `c_char`, NULL bulk-data with `len>0`, NULL covariance array, NULL descriptor string at build)
- LocalTime round-trip test hardened to exercise all 4 frame_id parities

**Verification:** `cargo test`, `cargo clippy --all-targets -- -D warnings`, `cargo fmt --all --check` — all clean.
